packages/core/src/i18n.t.ts

```diff
@@ -0,0 +1,148 @@
+  : never // If the string does not match the expected pattern, return 'never'.
+
+type ExtractFormatterMessages<Input extends string> = string extends Input
+  ? string[]
+  : Input extends ""
+  ? []
+  : Input extends `${string}{${infer Tail}`
+  ? ExtractBraceBody<Tail> extends [
+      infer BraceBody extends string,
+      infer Next extends string
+    ]
+    ? [BraceBody, ...ExtractFormatterMessages<Next>]
+    : []
+  : []
+
+type ExtractFormatterId<Input extends string> =
+  Input extends `${infer FormatterId},${string}`
+    ? Trim<FormatterId>
+    : Trim<Input>
+
+type IsExistedFormatter<FormatterId extends string> =
+  FormatterId extends keyof FormatterMap ? FormatterId : never
+
+type FormatterInputType<FormatterId extends string> =
+  FormatterMap[IsExistedFormatter<FormatterId>] extends infer FormatterFn
+    ? FormatterFn extends (input: infer InputType, ...args: any[]) => any
+      ? InputType
+      : never
+    : never
+
+/**
+ * Takes a string literal and extracts interpolation variables from it.
+ * @example `My name is {name}, i'm from {city}` -> {name: string} | {city: string}
+ * @return union of objects with one property each
+ */
+type ExtractVars<Input extends string> = string extends Input // Check if the type of Input is the general 'string' type and not a string literal.
+  ? Record<string, unknown> // If it is, return a wide type Record<string, unknown> to represent any object.
+  : // Start processing the literal string.
+  Input extends `${string}{${infer Tail}` // Check if the Input string contains '{', indicating the start of an expression.
+  ? // Extract the content inside the curly braces and the rest of the string after the closing curly brace.
+    ExtractBraceBody<Tail> extends [
+      infer BraceBody extends string,
+      infer Next extends string
+    ]
+    ? BraceBody extends `${infer FormatterInput},${infer FormatterTail}` // Check if the content inside the braces is a formatter.
+      ? // Process the formatter content.
+        | {
+              // Create a type with a property where the key is trimmed FormatterInput (which is var name in this case), and the value is determined by the formatter's type.
+              [P in Trim<FormatterInput>]: FormatterInputType<
+                ExtractFormatterId<FormatterTail>
+              >
+            }
+          | ExtractVars<ExtractFormatterMessages<FormatterTail>[number]> // Recursively process messages inside the formatter.
+          | ExtractVars<Next> // Recursively process the rest of the string after the current variable.
+      : // If not a formatter, create a type with a property where the key is the whole trimmed BraceBody and the value is a string
+        { [P in Trim<BraceBody>]: string } | ExtractVars<Next> // and recursively process the rest of the string.
+    : {} // If the string does not contain a valid structure, return an empty object type (possibly we can return an error here, because this branch indicates parsing error)
```

Comment:
`{}` does not represent an "empty object" in TS, so `ExtractVars` might be behaving differently from what you expect

see https://www.totaltypescript.com/the-empty-object-type-in-typescript

Reply:
Yes, thank you for pointing that out! I know about this specific issue, and it's actually the only place I also wasn't sure about :) In this context, it makes no difference because `ExtractVars` only applies if the input string contains interpolation. But in the general case, I should use `never` here, and it should be fixed, of course. Thank you again!